src/main/java/aquality/selenium/core/elements/ElementFinder.java

```diff
@@ -0,0 +1,77 @@
+package aquality.selenium.core.elements;
+
+import aquality.selenium.core.elements.interfaces.IElementFinder;
+import aquality.selenium.core.localization.ILocalizedLogger;
+import aquality.selenium.core.waitings.IConditionalWait;
+import com.google.inject.Inject;
+import org.openqa.selenium.By;
+import org.openqa.selenium.NoSuchElementException;
+import org.openqa.selenium.TimeoutException;
+import org.openqa.selenium.WebElement;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+/**
+ * Implementation of IElementFinder.
+ */
+public class ElementFinder implements IElementFinder {
+    private final ILocalizedLogger localizedLogger;
+    private final IConditionalWait conditionalWait;
+
+    @Inject
+    public ElementFinder(ILocalizedLogger localizedLogger, IConditionalWait conditionalWait) {
+        this.localizedLogger = localizedLogger;
+        this.conditionalWait = conditionalWait;
+    }
+
+    @Override
+    public List<WebElement> findElements(By locator, DesiredState desiredState, Long timeoutInSeconds) {
+        AtomicBoolean wasAnyElementFound = new AtomicBoolean(false);
+        List<WebElement> resultElements = new ArrayList<>();
+        try {
+            conditionalWait.waitFor(driver -> {
+                List<WebElement> currentFoundElements = driver == null ? new ArrayList<>() : driver.findElements(locator);
```

Comment:
where is the driver initialized? I am asking because trying to imagine what case driver might be null

Reply:
I don't think that the driver could possibly be null (it gained from the application class, from the DI). There is only one reason for this null check: IDE code-checking plugins suggest to add it